### PR TITLE
[Improve][Connector-V2][Iceberg] Add hadoop s3 catalog e2e testcase

### PR DESCRIPTION
**Pull request by 4chicat**
### Purpose of this pull request
We use S3 as iceberg warehouse storage.
If `catalog_type` is `hadoop`, only hdfs is supported, so it is being customized and used.
I want to contribute my customized code, so I push a pull request.

### Does this PR introduce _any_ user-facing change?
we can use the iceberg table saved in s3.

config example
```yml
source {
  iceberg {
    catalog_name = "seatunnel"
    iceberg.catalog.config={
      "type"="hadoop"
      "warehouse"="s3a://your_bucket/spark/warehouse/"
    }
    hadoop.config={
      "fs.s3a.path.style.access" = "true"
      "fs.s3a.connection.ssl.enabled" = "false"
      "fs.s3a.signing.algorithm" = "S3SignerType"
      "fs.s3a.encryption.algorithm" = "AES256"
      "fs.s3a.connection.timeout" = "3000"
      "fs.s3a.aws.credentials.provider" = "org.apache.hadoop.fs.s3a.SimpleAWSCredentialsProvider"
      "fs.s3a.endpoint" = "http://minio:9000"
      "fs.s3a.access.key" = "xxx"
      "fs.s3a.secret.key" = "xxx"
      "fs.defaultFS" = "s3a://xxx"
    }
    namespace = "your_iceberg_database"
    table = "your_iceberg_table"
    result_table_name = "iceberg_test"
  }
}
```

### How was this patch tested?
add e2e test case : `connector-iceberg-s3-e2e`

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by Hisoka-X** [comment]
cc @ic4y

**Review by EricJoy2048 — COMMENTED** [review]
LGTM, cc @ic4y

**Review by EricJoy2048 — DISMISSED** [review]

**Review by Hisoka-X — COMMENTED** [review]
Shall we add a test case for this feature?

**Comment by 4chicat** [comment]
> Shall we add a test case for this feature?

I don't think it's going to be easy because it requires an S3 bucket.

**Comment by Hisoka-X** [comment]
I think we can create a minio container in e2e test case for s3a file read.

**Comment by 4chicat** [comment]
@EricJoy2048 
Thank you so much for the review.  
I made an additional commit today by adding the e2e test case.
However, your previous review was automatically dismissed by the system.
I apologize for that.

**Comment by 4chicat** [comment]
> I think we can create a minio container in e2e test case for s3a file read.

I created minio test container and added e2e test case.
It passes on spark3, but fails on flink/seatunnel/spark2.

I think I need to add hadoop-aws/aws-java-sdk jar file to the test container docker image and need your help.
* spark2 : hadoop-aws-2.7.5, aws-java-sdk-1.7.4

I'm also looking to see if there are other workarounds.

**Comment by Hisoka-X** [comment]
Thanks @4chicat , I will take a look in the few days.

**Comment by 4chicat** [comment]
> Thanks @4chicat , I will take a look in the few days.

I modified to only run e2e tests for seatunnel engine and spark 3.3. (except flink and spark 2.4)

**Comment by EricJoy2048** [comment]
@Hisoka-X  @ic4y  PTAL. 
@Carl-Zhou-CN  @liugddx

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-s3-e2e/src/test/resources/log4j.properties
Comment (liugddx) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-s3-e2e/src/test/resources/log4j.properties:1]: Why add this file

Comment (4chicat) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-s3-e2e/src/test/resources/log4j.properties:1]: `connector-iceberg-s3-e2e` e2e case was written based on `connector-iceberg-hadoop3-e2e`.
This file was created by copying the file below.
https://github.com/apache/seatunnel/blob/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-hadoop3-e2e/src/test/resources/log4j.properties

Should I delete it?

Comment (liugddx) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-s3-e2e/src/test/resources/log4j.properties:1]: > `connector-iceberg-s3-e2e` e2e case was written based on `connector-iceberg-hadoop3-e2e`. This file was created by copying the file below. https://github.com/apache/seatunnel/blob/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-hadoop3-e2e/src/test/resources/log4j.properties
> 
> Should I delete it?

It is already included in the engine and can be removed

Comment (4chicat) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-s3-e2e/src/test/resources/log4j.properties:1]: > > `connector-iceberg-s3-e2e` e2e case was written based on `connector-iceberg-hadoop3-e2e`. This file was created by copying the file below. https://github.com/apache/seatunnel/blob/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-hadoop3-e2e/src/test/resources/log4j.properties
> > Should I delete it?
> 
> It is already included in the engine and can be removed

Removed `log4j.properties` file.

**Review by hailin0 — COMMENTED** [review]
LGTM

cc @Hisoka-X 
image

**Comment by Hisoka-X** [comment]
Look like this PR bring new leak thread in seatunnel server. Can you try closing these leaking threads? If not, please add it to the known issues list. https://github.com/apache/seatunnel/blob/92f847c99f219077b4d45c25d2b379d42879f613/seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/seatunnel/SeaTunnelContainer.java#L291

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/util/ContainerUtil.java [hunk 1]
```diff
@@ -251,6 +252,12 @@ public static List<TestContainer> discoverTestContainers() {
         }
     }
 
+    public static List<TestContainer> discoverTestSpark3Containers() {
+        return discoverTestContainers().stream()
+                .filter(testContainer -> testContainer instanceof Spark3Container)
+                .collect(Collectors.toList());
+    }
```
Comment (Hisoka-X) [inline seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/util/ContainerUtil.java]: please remove

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/seatunnel/SeaTunnelContainer.java [hunk 2]
```diff
@@ -228,6 +228,9 @@ private List<String> removeSystemThread(List<String> beforeThreads, List<String>
                                 || s.contains(
                                         "org.apache.hadoop.fs.FileSystem$Statistics$StatisticsDataReferenceCleaner")
                                 || s.startsWith("Log4j2-TF-")
+                                || s.contains("java-sdk-http-connection-reaper")
+                                || s.contains("Timer for 's3a-file-system' metrics system")
+                                || s.startsWith("MutableQuantiles-")
```
Comment (Hisoka-X) [inline seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/seatunnel/SeaTunnelContainer.java]: please move it into https://github.com/apache/seatunnel/pull/5745/files#diff-828bb95593582c2297e60ef9a6ced4733f251eef77389235d8c3b506dd0beb47L291

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/util/ContainerUtil.java (hunk 1 above)
Comment (4chicat) [reply seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/util/ContainerUtil.java]: It has been removed.

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/seatunnel/SeaTunnelContainer.java (hunk 2 above)
Comment (4chicat) [reply seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/seatunnel/SeaTunnelContainer.java]: Moved to `isIssueWeAlreadyKnow` method.

**Comment by EricJoy2048** [comment]
This PR is very valuable and I think we need to look at it again and review it.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-s3-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/s3/IcebergSourceIT.java [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.seatunnel.connectors.seatunnel.iceberg.config.IcebergCatalogType.HADOOP;
+
+@DisabledOnContainer(
+        value = {TestContainerId.SPARK_2_4},
+        type = {EngineType.FLINK},
+        disabledReason = "Needs hadoop-aws,aws-java-sdk jar for flink and spark2.4.")
```
Comment (Carl-Zhou-CN) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-s3-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/s3/IcebergSourceIT.java]: Could you please help me annotate this section in the usage document and provide an example for AWS S3？

Comment (4chicat) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-s3-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/s3/IcebergSourceIT.java]: Thanks for your comment. Added to document.

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/seatunnel/SeaTunnelContainer.java [hunk 4]
```diff
@@ -318,7 +318,11 @@ private boolean isIssueWeAlreadyKnow(String threadName) {
                 || threadName.startsWith("MaintenanceTimer")
                 || threadName.startsWith("cluster-")
                 // Iceberg
-                || threadName.startsWith("iceberg");
+                || threadName.startsWith("iceberg")
+                // Iceberg S3 Hadoop catalog
+                || threadName.contains("java-sdk-http-connection-reaper")
+                || threadName.contains("Timer for 's3a-file-system' metrics system")
+                || threadName.startsWith("MutableQuantiles-");
```
Comment (Carl-Zhou-CN) [inline seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/seatunnel/SeaTunnelContainer.java:433]: @Hisoka-X cc

File: seatunnel-shade/seatunnel-hadoop3-3.1.4-uber/pom.xml [hunk 5]
```diff
@@ -52,6 +52,11 @@
             <artifactId>snappy-java</artifactId>
             <version>1.1.10.4</version>
         </dependency>
+        <dependency>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-aws</artifactId>
+            <version>${hadoop3.version}</version>
+        </dependency>
```
Comment (Hisoka-X) [inline seatunnel-shade/seatunnel-hadoop3-3.1.4-uber/pom.xml]: cc @EricJoy2048

Comment (EricJoy2048) [reply seatunnel-shade/seatunnel-hadoop3-3.1.4-uber/pom.xml]: Why add `hadoop-aws` to hadoop shade jar?

**Comment by EricJoy2048** [comment]
> > I think we can create a minio container in e2e test case for s3a file read.
> 
> I created minio test container and added e2e test case. It passes on spark3, but fails on flink/seatunnel/spark2.
> 
> I think I need to add hadoop-aws/aws-java-sdk jar file to the test container docker image and need your help.
> 
> * spark2 : hadoop-aws-2.7.5, aws-java-sdk-1.7.4
> 
> I'm also looking to see if there are other workarounds.

Like the driver, you can add `hadoop-aws` and `aws-java-sdk` to the test container docker like `seatunnel-e2e/seatunnel-connector-v2-e2e/connector-file-oss-e2e/src/test/java/org/apache/seatunnel/e2e/connector/file/oss/OssFileIT.java`

```
@TestContainerExtension
    private final ContainerExtendedFactory extendedFactory =
            container -> {
                Container.ExecResult extraCommands =
                        container.execInContainer(
                                "bash",
                                "-c",
                                "mkdir -p /tmp/seatunnel/plugins/oss/lib && cd /tmp/seatunnel/plugins/oss/lib && curl -O "
                                        + OSS_SDK_DOWNLOAD);
                Assertions.assertEquals(0, extraCommands.getExitCode());

                extraCommands =
                        container.execInContainer(
                                "bash",
                                "-c",
                                "cd /tmp/seatunnel/plugins/oss/lib && curl -O " + JDOM_DOWNLOAD);
                Assertions.assertEquals(0, extraCommands.getExitCode());

                extraCommands =
                        container.execInContainer(
                                "bash",
                                "-c",
                                "cd /tmp/seatunnel/plugins/oss/lib && curl -O "
                                        + HADOOP_ALIYUN_DOWNLOAD);
                Assertions.assertEquals(0, extraCommands.getExitCode());

                extraCommands =
                        container.execInContainer(
                                "bash",
                                "-c",
                                "cd /tmp/seatunnel/lib && curl -O " + OSS_SDK_DOWNLOAD);
                Assertions.assertEquals(0, extraCommands.getExitCode());

                extraCommands =
                        container.execInContainer(
                                "bash", "-c", "cd /tmp/seatunnel/lib && curl -O " + JDOM_DOWNLOAD);
                Assertions.assertEquals(0, extraCommands.getExitCode());

                extraCommands =
                        container.execInContainer(
                                "bash",
                                "-c",
                                "cd /tmp/seatunnel/lib && curl -O " + HADOOP_ALIYUN_DOWNLOAD);
                Assertions.assertEquals(0, extraCommands.getExitCode());
            };
```

File: seatunnel-shade/seatunnel-hadoop3-3.1.4-uber/pom.xml (hunk 5 above)
Comment (4chicat) [reply seatunnel-shade/seatunnel-hadoop3-3.1.4-uber/pom.xml]: Changed to not add `hadoop-aws`.

**Review by EricJoy2048 — APPROVED** [review]

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/seatunnel/SeaTunnelContainer.java (hunk 4 above)
Comment (Hisoka-X) [reply seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/seatunnel/SeaTunnelContainer.java:433]: LGTM

**Review by hailin0 — APPROVED** [review]
LGTM

**Pull request merged**